Vfs with couchdb

jinroh opened the pull request:
This PR implement an almost complete API to upload files and create directories.

Still missing (indicated by `@TODO`s in the code):
  - a proper implementation of our jsonapi: this PR introduces a temporary `JSONApier` interface to do the job
  - ~~proper error handling to make the creation of a file/directory "atomic" to make sure out couchdb and VFS are coherent: this requires a `couchdb.Delete/Update`~~
  - ~~`Mime` and `Class` fields in file attributes (not sure how we want to handle that)~~

nono commented on web/files/file.go:
```diff
@@ -0,0 +1,151 @@
+	calcMD5 := md5H.Sum(nil)
+	if !bytes.Equal(m.GivenMD5, calcMD5) {
+		err = errInvalidHash
+		return
```
I think you can replace these 2 lines by this one:

```go
return errInvalidHash
```

jinroh replied:
done fbb9fb1c850b6b3b4b984e3b9e960d3d5de3270c

nono commented on web/files/file.go:
```diff
@@ -0,0 +1,151 @@
+		} else {
+			fs.Remove(pth)
+		}
+	}()
```
This error handler looks weird to me. I'd have simply use that:

```go
defer func() {
  if err != nil {
    fs.Remove(pth)
  }
}
```

What am I missing?

jinroh commented on web/files/directory.go:
```diff
@@ -0,0 +1,111 @@
+}
+
+// CreateDirectory is the method for creating a new directory
+func CreateDirectory(m *DocMetadata, fs afero.Fs, dbPrefix string) (jsonapier jsonapi.JSONApier, err error) {
```
This is what i did initially, and gometalinter hinted that I was returning a private struct on a public func.

I didn't really think it is necessary to make dirDoc a public struct. I could also make createDirectory a private func for now.

nono replied:
Well, I think that making `dirDoc` public may be the good choice. We want other packages to use `CreateDirectory` and it make senses that they know what is a `dirDoc` to use the returned value.

jinroh replied:
Let's go for it. I'll do the same for `fileDoc`.

jinroh replied:
done 83f36d4043d3247b2ac31b354bd64af89a434ebb

jinroh commented on web/files/files_test.go:
```diff
@@ -9,10 +12,13 @@ import (
 	"github.com/stretchr/testify/assert"
 )
 
+const COUCHURL = "http://localhost:5984/"
```
copypasta. done d72690e147da9743ead284cd008ed97b72dd69c9

nono commented:
> `Mime` and `Class` fields in file attributes (not sure how we want to handle that)

`Mime` is the `Content-Type` HTTP header in the request (you can use `application/octet-stream` for the default value).

`Class` is computed from `Mime`. In the actual cozy, each application has its own rules for setting the `class`. https://github.com/cozy/cozy-files/blob/ea9f31c772c7002ef5853361ba8ffe6a6942b5b3/server/helpers/file.coffee#L18-L25 for example. For the moment, let's say we use the first part of the `Mime` (`image`, `audio`, `video`, `text`, etc.) and we will improve this for specific mime-type later.

nono commented on web/files/file.go:
```diff
@@ -163,3 +167,14 @@ func copyOnFsAndCheckIntegrity(m *DocMetadata, fs afero.Fs, pth string, r io.Rea
+
+	// @TODO improve for specific mime types
+	mime = contentType
+	class = contentType[:strings.Index(contentType, "/")]
```
```
class = strings.Split(contentType, "/")[1]
```

is more explicit, no?

nono replied:
it's OK for me

jinroh commented:
I've just noted that I missed the file size. We should probably handle the `Content-Length` and return an error on:

  - length mismatch (contentlength != copied)
  - content-length not specified

nono commented:
> length mismatch (contentlength != copied)

Yes!

> content-length not specified

We can let `content-length` be optional.

jinroh commented:
Added content length mismatch support (22c2cd5b61605c5552b180400078b914df51da1a) and removed charset from mime (3022445f0a9b135138cdba8f031180412a394ab0).
